### PR TITLE
Updated patterns and added generation script

### PR DESCRIPTION
**Pull request by Yethal**
Updated patterns to include all commands present in 0.63.0
Added nu script that will update the file automatically

### CODE REVIEW

**Comment by fdncred** [comment]
cool, a contribution! Thanks!

We need to test this pretty thoroughly. I suspect some of these regex changes won't work right (I could be wrong). In some circumstances I put the subs first because things didn't work right without them being first.

I don't really like the formatting changes because it makes it harder to read, although I could be talked into changing the format.

**Comment by fdncred** [comment]
What would really be nice is some test script in the repo with *EVERY* command in it to make sure we're highlighting it correctly in context. The script doesn't have to be runnable though. I'm not quite sure. Just something to test with.

**Comment by Yethal** [comment]
Formatting was done by nushell, not by me so I can't really help with that, maybe adding prettier as a hook in husky would help with that?
I'll add the test script and fix for question marsk tomorrow (it's middle of the night here)

**Comment by Yethal** [comment]
I added script to generate the example script, let me know if that's enough

**Comment by fdncred** [comment]
oh, wow. that's a good idea to use examples. nice job.  BTW, [this](https://github.com/nushell/nu_scripts/blob/main/make_release/gen-js-ext.nu) is the script I used to use to generate the regexes. I'm not saying we have to use it. yours are probably fine.

I need to test this on my windows box. will probably happen next week sometime since I'm on vacation this week.

**Comment by fdncred** [comment]
I'm starting to remember that the subcommands have to come first or else the syntax highlighting doesn't work right. so, like `ansi gradient` needs to be declared before `ansi`, which is why I had all the subs first. I'll know for sure once I test.

**Comment by fdncred** [comment]
I'm remembering something else. Maybe you can help me figure this out too. You'll notice that my previous regexes don't have all the keywords, for instance `def` is not in the `keywords.d`. That is because if I put it in a keyword, I can't recognize it when it's actually used in `def blah [] {}` and highlight everything.

I really want to figure out how to use your scripts and just make this entirely autogenerated.

**Comment by Yethal** [comment]
I'm not entirely sure what the problem is but looking at completions for other languages it seems the regexes are supposed to use capture groups to separate tokens on the same line that belong to different categories (such as def and blah being a keyword and function name). Take a look at ruby's file: https://github.com/rubyide/vscode-ruby/blob/main/packages/vscode-ruby/syntaxes/ruby.cson.json

**Comment by fdncred** [comment]
ya, I've looked at a lot of these type of files via cs.github.com seeing how other people do it. there are apparently many ways to write this mess.

**Comment by Yethal** [comment]
@fdncred I added some scripts to simplify local testing, now after changing the syntax file run npm run package and reload vscode window and newer highlighting will be autoloaded.

**Comment by fdncred** [comment]
I hope to find some time to look at this in more detail and try it out on my windows box where I developed the extension. I'm still concerned about the points I've made above and think that if we just apply these changes it won't work right, but I want to see if I'm right or wrong. I sure appreciate you looking into this though. :)

File: package.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     "webpack-dev": "webpack --mode development --watch",
     "test-compile": "tsc -p ./",
     "package": "vsce package",
-    "postpackage": "./install-extension.nu"
+    "postpackage": "code --install-extension *.vsix"
```
[the remark this answers is not in the document]
Comment (Yethal) [reply package.json:35]: If it'd help I can make it load the newest vsix file

**Comment by fdncred** [comment]
I was looking at this today. Some things aren't highlighted correctly. I'm not saying these have ever been right necessarily. I'm sure some of the problems are what I mention above but others perhaps have never worked right.

Just documenting here.

generally speaking, any command/sub-command with a dash `-` in the name is colored wrong.

`date to-timezone`
image

`decode utf-8`
image

`ls` parameters
image

`dfr to-df`, `dfr all-false`, `dfr all-true`, `dfr arg-sort`
image

Would be nice to color module names and use parameter better
image

This `.com` should be colored the same as `url`
image

`group-by`
image

also variable names with dashes
image

`sort-by`
image

`overlay` sub-commands and names would be nice to color differently
image

`path relative-to`
image

`skip until` and `skip while` - i these are colored wrong because `skip` is probably defined first
image

same scenario with `split`
image

`str` commands need work
image

`take` is similar to the `skip` and `split` commands
image

`to` sub commands aren't colored correctly either
image

`url` sub-commands - notich `path` and `query` look good but that's because it thinks they are their own commands, which they are
image

**Comment by Yethal** [comment]
I think I figured it out. For two word commands we need to put only the second words in alternation group with greed like this:
\bansi(\sgradient|\sstrip)?\b

**Comment by fdncred** [comment]
there appear to be a lot of single terms like this
```json
       {
           "name": "keyword.other.a",
           "match": "\\balias"
         },
         {
           "name": "keyword.other.a",
           "match": "\\ball-false"
         },
         {
           "name": "keyword.other.a",
           "match": "\\ball-true"
         },
         {
           "name": "keyword.other.a",
           "match": "\\ball\\?"
         },
```
I think we need to combine what we can instead of having hundreds of keyword.other.?

**Comment by Yethal** [comment]
Try newest revision, I'm now combining all single word commands into a single match

**Comment by fdncred** [comment]
I think it's closer but not quite right yet.

the first `keyword.other.a` should be `keyword.other.a_sub`.
the second `keyword.other.a` looks like it wants to be a a_sub but there's nothing in the `()` so it probably belongs with the dash ones in the third `keyword.other.a`.

this seems to repeat for most letters.
```json
        {
          "name": "keyword.other.a",
          "match": "\\bansi(\\s|\\sgradient|\\sstrip)?\\b"
        },
        {
          "name": "keyword.other.a",
          "match": "\\bappend(\\s|\\s)?\\b"
        },
        {
          "name": "keyword.other.a",
          "match": "\\b(agg|agg-groups|alias|all-false|all-true|all\\?|any\\?|arg-max|arg-min|arg-sort|arg-true|arg-unique|as|as-date|as-datetime)\\b"
        },
        ```

**Comment by fdncred** [comment]
This looks better but we're still highlighting things wrong. For instance, `to-df`, which doesn't exist any longer but the point is still valid here.
image
You can see that `to` is recognized but not `to-df`
image
I believe this is because `to` is listed in `keyword.other.t` and maybe it shouldn't
image
If i change it to this:
image
The highlighting looks like this:
image
and you can see that `to-df` is not recognized
image

**Comment by fdncred** [comment]
i regenerated the example.nu file and some things look good, others do not.

`all-false` and `all-true` look great. `into` looks good `df` does not
image
`ansi gradient` and `ansi strip` not good
image
`date humanize` `date list-timezone` `date now` not good
image
`drop-duplicates`
image
most of the `into blahs`
image
some of the `math` subcommands look ok but i'm betting this is because `max` `median` and `min` are keywords somewhere on their own
image
same with `path blahs` and `random blahs` and `roll blahs` and `skip blahs` and `split blahs` basicially anything with a `command<space>subcommand`
image

if we can fix these few things, we can probably get to a new release. I think it's really about ordering, i.e. the sub commands need to come first and things like i mentioned about with `to`.

**Comment by fdncred** [comment]
@Yethal What are the odds of working more on this? I would be nice to publish a new version.

**Comment by Yethal** [comment]
I still want to work on this, just didn't have time recently. I'll experiment with more strict capture groups today

**Comment by fdncred** [comment]
great! I thought I had made you mad with my comments about things not being right/fixed. :)

**Comment by Yethal** [comment]
I am kinda mad about this but not at you, rather at vscode and it overcomplicated syntax highlighting system. Who would've thought that this much effort is needed to achieve something so simple

**Comment by fdncred** [comment]
agreed, i yell at vscode every time i have to update it, which is why i was so happy when you were writing a automation script. :)

I really think we can get it right but the keyword regular expressions just have to be in a particular order in the file.

at a later time, i'd like to add some typescript stuff to utilize the new `nu-check` command. but that's for another pr.

**Comment by fdncred** [comment]
@Yethal I haven't tried your latest commit. Do you think it's working now?

**Comment by Yethal** [comment]
I scrolled through the generated example and it looks good now

**Comment by fdncred** [comment]
Wow! All in all this is 100% better. There are just a few outliers that we need to fix. I think most of these are the same problem, for instance `group`'s regex is defined before `group-by`'s regex. The fix is to put `group-by` in a regex definition before `group`. This is how I fixed it.
```json
        {
          "name": "keyword.other.g",
          "match": "\\b(group-by)\\w*"
        },
        {
          "name": "keyword.other.g",
          "match": "\\b(g|get|get-day|get-hour|get-minute|get-month|get-nanosecond|get-ordinal|get-second|get-week|get-weekday|get-year|glob|grid|group|gstat)\\w*"
        },

```
fixed group-by
image

## Outliers we still need to fix
group-by
image

each while
image

encode base64
image

format filesize
image

nu-check
image

set-with-idx
image

**Comment by fdncred** [comment]
Let's land this with the manual tweaks so we can do a release with the latest syntax and circle back for more changes to generate-patterns.nu.

**Pull request merged**